zephyr/Kconfig

```diff
@@ -0,0 +1,680 @@
+# Copyright (c) 2017-2020 Linaro Limited
+# Copyright (c) 2020 Arm Limited
+# Copyright (c) 2022 Legrand North America, LLC.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+
+# Option to build the project with the MCUBoot application
+config MCUBOOT
+	default n
+	bool  "MCUBoot interfaces"
+	select MCUBOOT_APP
+
+config MCUBOOT_APP
+	default n
+	bool  "MCUBoot application"
```

Comment:
Why is this selectable?

Reply:
Project configurations outside the mcuboot repository need a Kconfig to control whether or not the MCUBoot application is built into the executable.  MCUBOOT_APP provides that control.

Reply:
Projects in other repositories need to be able to configure whether or not the MCUboot application is built into their image.  This is the primary use case that motivated this refactoring.

Reply:
I **must** be able to build and flash the boot code for release independent of Sysbuild.  Running boot code without a valid image is also a necessary test case. Sysbuild is an efficiency enhancement, not a base case of usage.

Reply:
> Projects in other repositories need to be able to configure whether or not the MCUboot application is built into their image. This is the primary use case that motivated this refactoring.

This is just wrong, mcuboot is a immutable bootloader, it is separate from the application and should not be combined, the point in mcuboot is it's small and does as little as possible to expose the least possible attack surface to ensure root of trust. If you're building it into your application, you might as well just disable security entirely at that point.

Reply:
> mcuboot is a immutable bootloader

That is one use case. Another necessary one is for mcuboot to be a secondary (and mutable) bootloader - needed when the immutable block of the device is too small for the full bootloader functionality.

> it is separate from the application and should not be combined,

You cannot logically claim that there is only one application, that it is separate from the bootloader, AND that the bootloader is an application.

> the point in mcuboot is it's small and does as little as possible to expose the least possible attack surface to ensure root of trust.

This description is not unique to mcuboot. It is also true for many certifiably secure firmware images.  The unique point of bootloaders is to load and execute another image. Doing "as little as possible" varies based on user needs.  Some need serial recovery. Some need it to run from ram. Some need it to support multiple partitions.  Some need it to prevent version downgrades. Each feature combination could be supported by unique source files, but it is more efficient to support them with reusable, configurable code. Using Zephyr's definition of "application", each configuration combination to build mcuboot is a distinct application in its own right.

BTW, it is actually very reasonable to implement outputing/logging "hello_world" as a trivial subsystem launched by `SYS_INIT()`, and integrate this into a bootloader. It makes even more sense if this subsystem existed in within a proprietary module and logged the executable version rather than the string "hello world". ;-)